File: src/test/java/org/stt/gui/jfx/binding/ReportBindingTest.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package org.stt.gui.jfx.binding;

import javafx.beans.property.SimpleObjectProperty;
import org.hamcrest.CoreMatchers;
import org.hamcrest.collection.IsEmptyCollection;
import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.stt.model.TimeTrackingItem;
import org.stt.persistence.ItemReader;
import org.stt.query.TimeTrackingItemQueries;
import org.stt.reporting.SummingReportGenerator;

import javax.inject.Provider;
import java.time.Duration;
import java.time.LocalDate;
import java.util.Optional;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.not;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.mock;

/**
 *
 * @author dante
 */
public class ReportBindingTest {

	private ReportBinding sut;
    private SimpleObjectProperty<LocalDate> reportStart = new SimpleObjectProperty<>();
    private SimpleObjectProperty<LocalDate> reportEnd = new SimpleObjectProperty<>();

    private Provider<ItemReader> readerProvider;
    private ItemReader itemReader = mock(ItemReader.class);

	@Before
	public void setup() {
        readerProvider = () -> itemReader;
        sut = new ReportBinding(reportStart, reportEnd, new TimeTrackingItemQueries(readerProvider, Optional.empty()));
    }

	@Test
	public void shouldReturnEmptyReportIfStartIsNull() {
		// GIVEN
        reportEnd.set(LocalDate.now());

		// WHEN
		SummingReportGenerator.Report result = sut.getValue();

		// THEN
		Assert.assertThat(result.getStart(), CoreMatchers.nullValue());
		Assert.assertThat(result.getEnd(), CoreMatchers.nullValue());
		Assert.assertThat(result.getReportingItems(), IsEmptyCollection.empty());
		Assert.assertThat(result.getUncoveredDuration(), is(Duration.ZERO));
	}

	@Test
	public void shouldReturnEmptyReportIfEndIsNull() {
		// GIVEN
        reportStart.set(LocalDate.now());

		// WHEN
		SummingReportGenerator.Report result = sut.getValue();

		// THEN
		Assert.assertThat(result.getStart(), CoreMatchers.nullValue());
		Assert.assertThat(result.getEnd(), CoreMatchers.nullValue());
		Assert.assertThat(result.getReportingItems(), IsEmptyCollection.empty());
		Assert.assertThat(result.getUncoveredDuration(), is(Duration.ZERO));
	}

	@Test
	public void shouldReadItemsIfStartAndEndAreSet() {
		// GIVEN
        final LocalDate start = LocalDate.now();
        reportStart.set(start);
        final LocalDate end = start.plusDays(1);
        reportEnd.set(end);
        TimeTrackingItem item = new TimeTrackingItem("none", start.atStartOfDay(), end.atStartOfDay());
        given(itemReader.read()).willReturn(Optional.of(item), Optional.empty());

		// WHEN
		SummingReportGenerator.Report result = sut.getValue();

		// THEN
        Assert.assertThat(result.getStart(), is(start.atStartOfDay()));
        Assert.assertThat(result.getEnd(), is(end.atStartOfDay()));
        Assert.assertThat(result.getReportingItems(), not(IsEmptyCollection.empty()));
		Assert.assertThat(result.getUncoveredDuration(), is(Duration.ZERO));
	}
}

File: src/test/kotlin/org/stt/gui/jfx/binding/ReportBindingTest.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package org.stt.gui.jfx.binding

import javafx.beans.property.SimpleObjectProperty
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.mockito.BDDMockito.given
import org.mockito.Mockito.mock
import org.stt.model.TimeTrackingItem
import org.stt.persistence.ItemReader
import org.stt.query.TimeTrackingItemQueries
import java.time.Duration
import java.time.LocalDate
import java.util.*
import javax.inject.Provider

/**
 *
 * @author dante
 */
class ReportBindingTest {

    private var sut: ReportBinding? = null
    private val reportStart = SimpleObjectProperty<LocalDate>()
    private val reportEnd = SimpleObjectProperty<LocalDate>()

    private var readerProvider: Provider<ItemReader>? = null
    private val itemReader = mock(ItemReader::class.java)

    @Before
    fun setup() {
        readerProvider = Provider { itemReader }
        sut = ReportBinding(reportStart, reportEnd,
                TimeTrackingItemQueries(readerProvider!!, Optional.empty()))
    }

    @Test
    fun shouldReturnEmptyReportIfStartIsNull() {
        // GIVEN
        reportEnd.set(LocalDate.now())

        // WHEN
        val result = sut!!.value

        // THEN
        assertThat(result.start).isNull()
        assertThat(result.end).isNull()
        assertThat(result.reportingItems).isEmpty()
        assertThat(result.uncoveredDuration).isEqualTo(Duration.ZERO)
    }

    @Test
    fun shouldReturnEmptyReportIfEndIsNull() {
        // GIVEN
        reportStart.set(LocalDate.now())

        // WHEN
        val result = sut!!.value

        // THEN
        assertThat(result.start).isNull()
        assertThat(result.end).isNull()
        assertThat(result.reportingItems).isEmpty()
        assertThat(result.uncoveredDuration).isEqualTo(Duration.ZERO)
    }

    @Test
    fun shouldReadItemsIfStartAndEndAreSet() {
        // GIVEN
        val start = LocalDate.now()
        reportStart.set(start)
        val end = start.plusDays(1)
        reportEnd.set(end)
        val item = TimeTrackingItem("none", start.atStartOfDay(), end.atStartOfDay())
        given<TimeTrackingItem>(itemReader.read()).willReturn(item, null)

        // WHEN
        val result = sut!!.value

        // THEN
        assertThat(result.start).isEqualTo(start.atStartOfDay())
        assertThat(result.end).isEqualTo(end.atStartOfDay())
        assertThat(result.reportingItems).isNotEmpty
        assertThat(result.uncoveredDuration).isEqualTo(Duration.ZERO)
    }
}
